L1Trigger/Phase2L1ParticleFlow/plugins/L1HPSPFTauProducer.cc

```diff
@@ -0,0 +1,178 @@
+#include <vector>
+#include <numeric>
+
+////////////////////
+// FRAMEWORK HEADERS
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/L1TParticleFlow/interface/PFCandidate.h"
+#include "DataFormats/L1TParticleFlow/interface/PFTau.h"
+#include "DataFormats/JetReco/interface/CaloJet.h"
+// bitwise emulation headers
+#include "L1Trigger/Phase2L1ParticleFlow/interface/taus/L1HPSPFTauEmulator.h"
+
+class L1HPSPFTauProducer : public edm::global::EDProducer<> {
+public:
+  explicit L1HPSPFTauProducer(const edm::ParameterSet&);
+  ~L1HPSPFTauProducer() override{};
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+
+  //various needed vars
+  int nTaus_;
+  bool HW_;
+  bool fUseJets_;
+  bool debug_;
+
+  edm::EDGetTokenT<l1t::PFCandidateCollection> tokenL1PFCands_;
+  //jets
+  edm::EDGetTokenT<std::vector<reco::CaloJet>> tokenL1PFJets_;
+  //functions
+  std::vector<l1t::PFTau> processEvent_HW(std::vector<edm::Ptr<l1t::PFCandidate>>& parts,
+                                          std::vector<edm::Ptr<reco::CaloJet>>& jets) const;
+
+  static std::vector<l1HPSPFTauEmu::Particle> convertJetsToHW(std::vector<edm::Ptr<reco::CaloJet>>& edmJets);
+  static std::vector<l1HPSPFTauEmu::Particle> convertEDMToHW(std::vector<edm::Ptr<l1t::PFCandidate>>& edmParticles);
+  static std::vector<l1t::PFTau> convertHWToEDM(std::vector<l1HPSPFTauEmu::Tau> hwTaus);
+  //ADD
+  edm::EDPutTokenT<l1t::PFTauCollection> tauToken_;
+};
+
+L1HPSPFTauProducer::L1HPSPFTauProducer(const edm::ParameterSet& cfg)
+    : nTaus_(cfg.getParameter<int>("nTaus")),
+      HW_(cfg.getParameter<bool>("HW")),
+      fUseJets_(cfg.getParameter<bool>("useJets")),
+      debug_(cfg.getParameter<bool>("debug")),
+      tokenL1PFCands_(consumes(cfg.getParameter<edm::InputTag>("srcL1PFCands"))),
+      tokenL1PFJets_(consumes(cfg.getParameter<edm::InputTag>("srcL1PFJets"))),
+      tauToken_{produces()} {}
+
+void L1HPSPFTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("srcL1PFCands", edm::InputTag("l1tLayer1", "Puppi"));
+  desc.add<int>("nTaus", 16);
+  desc.add<bool>("HW", true);
+  desc.add<bool>("useJets", false);
+  desc.add<bool>("debug", false);
+  desc.add<edm::InputTag>("srcL1PFJets",
+                          edm::InputTag("l1tPhase1JetCalibrator9x9trimmed", "Phase1L1TJetFromPfCandidates"));
+  descriptions.add("L1HPSPFTauProducer", desc);
```

Comment:
Changing this to
```suggestion
  descriptions.add("l1HPSPFTauProducer", desc);
```
or
```suggestion
  descriptions.addWithDefaultLabel(desc);
```
would have kept the module label to start with the lower case letter, which would be more consistent with the other `lt1*` data products kept in the event content. (it would have changed the generated `cfi` file name though)

Reply:
Makes sense, the name formatting does stand out. Is this just something to keep in mind, or a change that should be made now (and if so, which of the two options is preferable)?

Reply:
I'd encourage to make things more consistent now (this kind of changes tend become more costly later), but it is really up to @cms-sw/l1-l2 and @cms-sw/upgrade-l2.

As of which one to prefer, it is somewhat a matter of taste (in this case). The `addWithDefaultLabel()` would be shorter, and would lead to less repetition of the class name.

Reply:
@makortel is right in this case, it is technically our naming scheme and one of those things that tend to both be easily forgotten and an annoyance. If you can make the change I think we can re-sign.

Reply:
I've gone ahead and done this in the latest update, no other changes.